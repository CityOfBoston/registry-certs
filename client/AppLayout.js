// @flow
/* eslint react/display-name: 0 */

import React, {
  type Element as ReactElement,
  type ChildrenArray as ReactChildrenArray,
} from 'react';

import headerHtml from '../templates/header.html';
import navigationHtml from '../templates/navigation.html';
import footerHtml from '../templates/footer.html';

import type { ClientDependencies } from './app';
import Nav, { type Props as NavProps, type LinkOptions } from './common/Nav';

type Props = {|
  children: ReactChildrenArray<ReactElement<*>>,
  navProps: NavProps,
|};

export default function AppLayout({ children, navProps }: Props) {
  // TODO(fin): remove wrapper <div> and return an array w/ React 16
  return (
    <div>
      <input
        type="checkbox"
        id="brg-tr"
        className="brg-tr"
        aria-hidden="true"
      />
      <nav
        className="nv-m"
        dangerouslySetInnerHTML={{ __html: navigationHtml }}
      />

      <div className="a11y--h" aria-live="polite" id="ariaLive" />

      <div className="mn mn--full mn--nv-s" style={{ zIndex: 2 }}>
        <input type="checkbox" id="s-tr" className="s-tr" aria-hidden="true" />
        <header
          className="h"
          role="banner"
          dangerouslySetInnerHTML={{ __html: headerHtml }}
        />

        <Nav {...navProps} />

        {children}
      </div>

      <footer
        className="ft"
        style={{ position: 'relative', zIndex: 2 }}
        dangerouslySetInnerHTML={{ __html: footerHtml }}
      />
    </div>
  );
}

export const wrapAppLayout = <P>(
  navLink: LinkOptions,
  render: (ClientDependencies, P) => ReactElement<*>
) => (dependencies: ClientDependencies, props: P) => (
  <AppLayout navProps={{ cart: dependencies.cart, link: navLink }}>
    {render(dependencies, props)}
  </AppLayout>
);
