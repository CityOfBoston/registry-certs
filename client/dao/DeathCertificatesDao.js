// @flow

import DataLoader from 'dataloader';
import type { LoopbackGraphql } from '../lib/loopback-graphql';
import type { DeathCertificate } from '../types';

import fetchDeathCertificates from '../queries/fetch-death-certificates';
import searchDeathCertificates from '../queries/search-death-certificates';
import lookupDeathCertificateOrder from '../queries/lookup-death-certificate-order';
import type {
  DeathCertificateSearchResults,
  DeathCertificateOrder,
} from '../types';

export type DeathCertificateCache = { [id: string]: DeathCertificate };

export default class DeathCertificatesDao {
  loopbackGraphql: LoopbackGraphql;
  loader: DataLoader<string, ?DeathCertificate>;

  cacheFullQuery: string;
  cacheByPage: { [page: number]: DeathCertificateSearchResults };

  constructor(loopbackGraphql: LoopbackGraphql) {
    this.loopbackGraphql = loopbackGraphql;

    // create new array shenanigans to get Flow to accept that we're not returning Errors
    this.loader = new DataLoader(ids =>
      fetchDeathCertificates(loopbackGraphql, ids).then(a => [...a])
    );
  }

  get(id: string): Promise<?DeathCertificate> {
    return this.loader.load(id);
  }

  async search(
    fullQuery: string,
    page: number
  ): Promise<DeathCertificateSearchResults> {
    if (fullQuery === this.cacheFullQuery) {
      const results = this.cacheByPage[page];
      if (results) {
        return results;
      }
    } else {
      this.cacheFullQuery = fullQuery;
      this.cacheByPage = {};
    }

    const { query, startYear, endYear } = this.parseQuery(fullQuery);

    const results = await searchDeathCertificates(
      this.loopbackGraphql,
      query,
      page,
      startYear,
      endYear
    );

    this.cacheByPage[page] = results;

    results.results.forEach(cert => {
      this.loader.prime(cert.id, cert);
    });

    return results;
  }

  async lookupOrder(
    id: string,
    contactEmail: string
  ): Promise<?DeathCertificateOrder> {
    return lookupDeathCertificateOrder(this.loopbackGraphql, id, contactEmail);
  }

  parseQuery(
    fullQuery: string
  ): { query: string, startYear: ?string, endYear: ?string } {
    // match a 4-digit year, and optionally a second 4-digit year with a hypen or en-dash
    const yearRegexp = /(\d{4})\s*[-â€“]?\s*(\d{4})?/;

    let query;
    let startYear;
    let endYear;

    const match = fullQuery.match(yearRegexp);
    if (match) {
      query = fullQuery.replace(yearRegexp, '').trim();
      startYear = match[1];
      endYear = match[2] || match[1];
    } else {
      query = fullQuery;
      startYear = null;
      endYear = null;
    }

    return {
      query,
      startYear,
      endYear,
    };
  }
}
