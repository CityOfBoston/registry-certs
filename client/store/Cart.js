// @flow

import { observable, computed, action, autorun } from 'mobx';

import type { DeathCertificate } from '../types';
import type DeathCertificatesDao from '../dao/DeathCertificatesDao';

type LocalStorageEntry = {|
  id: string,
  quantity: number,
|};

export class CartEntry {
  id: string;
  @observable.ref cert: ?DeathCertificate;
  @observable quantity: number;
}

export default class Cart {
  @observable entries: Array<CartEntry> = [];
  @observable pendingFetches: number = 0;

  localStorageDisposer: ?Function;

  attach(localStorage: Storage, deathCertificatesDao: DeathCertificatesDao) {
    if (localStorage) {
      try {
        const savedCart: Array<LocalStorageEntry> = JSON.parse(
          localStorage.getItem('cart') || '[]'
        );

        this.entries = savedCart.filter(({ quantity }) => quantity > 0).map(
          action(
            'hydrate entry from local storage start',
            ({ id, quantity }: LocalStorageEntry) => {
              const entry = new CartEntry();
              entry.id = id;
              entry.cert = null;
              entry.quantity = quantity;

              this.pendingFetches += 1;

              deathCertificatesDao.get(id).then(
                action(
                  'hydrate item from local storage complete',
                  (cert: ?DeathCertificate) => {
                    if (cert) {
                      entry.cert = cert;
                    } else {
                      this.remove(id);
                    }
                    this.pendingFetches -= 1;
                  }
                )
              );

              return entry;
            }
          )
        );
      } catch (e) {
        localStorage.removeItem('cart');
      }

      this.localStorageDisposer = autorun('save cart to local storage', () => {
        localStorage.setItem(
          'cart',
          JSON.stringify(
            this.entries.map(({ id, quantity }): LocalStorageEntry => ({
              id,
              quantity,
            }))
          )
        );
      });
    }
  }

  detach() {
    if (this.localStorageDisposer) {
      this.localStorageDisposer();
      this.localStorageDisposer = null;
    }
  }

  @computed
  get size(): number {
    // quantity shouldn't be below 0 but we want to be defensive.
    return this.entries.reduce(
      (acc, item) => acc + Math.max(item.quantity, 0),
      0
    );
  }

  @computed
  get loading(): boolean {
    return this.pendingFetches > 0;
  }

  add(cert: DeathCertificate, quantity: number) {
    const existingItem = this.entries.find(item => item.id === cert.id);
    if (existingItem) {
      existingItem.quantity += quantity;
    } else {
      const item = new CartEntry();
      item.id = cert.id;
      item.cert = cert;
      item.quantity = quantity;

      this.entries.push(item);
    }
  }

  setQuantity(cert: DeathCertificate, quantity: number) {
    const existingItem = this.entries.find(({ id }) => id === cert.id);
    if (existingItem) {
      // We don't remove items here when their quantity is 0 so that they don't
      // disappear when you edit the values on the cart page.
      existingItem.quantity = quantity;
    } else {
      this.add(cert, quantity);
    }
  }

  remove(certId: string) {
    const idx = this.entries.findIndex(({ id }) => id === certId);

    if (idx !== -1) {
      this.entries.splice(idx, 1);
    }
  }

  getQuantity(certId: string): number {
    const entry = this.entries.find(({ id }) => id === certId);

    if (entry) {
      return entry.quantity;
    } else {
      return 0;
    }
  }

  clean() {
    this.entries = this.entries.filter(({ quantity }) => quantity > 0);
  }

  clear() {
    this.entries = [];
  }

  @computed
  get containsPending(): boolean {
    return !!this.entries.find(({ cert }) => cert && cert.pending);
  }
}
