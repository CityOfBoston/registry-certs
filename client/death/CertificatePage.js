// @flow

import React, { type Element as ReactElement } from 'react';
import Head from 'next/head';

import type { DeathCertificate } from '../types';

import {
  getDependencies,
  type ClientContext,
  type ClientDependencies,
} from '../app';

import { wrapAppLayout } from '../AppLayout';

type InitialProps = {|
  id: string,
  certificate: ?DeathCertificate,
|};

type ContentProps = {
  ...InitialProps,
  addToCart: number => mixed,
};

type ContentState = {
  quantity: number,
};

export class CertificatePageContent extends React.Component<
  ContentProps,
  ContentState
> {
  state = {
    quantity: 1,
  };

  handleQuantityChange = (ev: SyntheticInputEvent<*>) => {
    this.setState({
      quantity: parseInt(ev.target.value, 10),
    });
  };

  handleAddToCart = (ev: SyntheticInputEvent<*>) => {
    ev.preventDefault();

    const { addToCart } = this.props;
    const { quantity } = this.state;

    addToCart(quantity);
  };

  render() {
    const { id, certificate } = this.props;

    return (
      <div>
        <Head>
          <title>Boston.gov — Death Certificate #{id}</title>
        </Head>

        <div className="p-a300">
          <div className="sh sh--b0">
            <h1 className="sh-title" style={{ marginBottom: 0 }}>
              Deceased Details
            </h1>
          </div>
        </div>

        <div className="p-a300 b--w">
          {certificate && this.renderCertificate(certificate)}
        </div>
      </div>
    );
  }

  renderCertificate({
    firstName,
    lastName,
    age,
    deathDate,
    deathYear,
  }: DeathCertificate) {
    const { quantity } = this.state;

    return (
      <div>
        <ul className="dl">
          <li className="dl-i">
            <span className="dl-t">Full Name</span>
            <span className="dl-d">
              {firstName} {lastName}
            </span>
          </li>
          <li className="dl-i">
            <span className="dl-t">Date of Death</span>
            <span className="dl-d">{deathDate || deathYear}</span>
          </li>
          <li className="dl-i">
            <span className="dl-t">Age</span>
            <span className="dl-d">{age}</span>
          </li>
        </ul>

        <form
          onSubmit={this.handleAddToCart}
          className="js-add-to-cart-form m-v300"
        >
          <select
            name="quantity"
            value={quantity}
            className="quantity"
            onChange={this.handleQuantityChange}
          >
            <option value="1">Qty: 1</option>
            <option value="2">Qty: 2</option>
            <option value="3">Qty: 3</option>
            <option value="4">Qty: 4</option>
            <option value="5">Qty: 5</option>
            <option value="6">Qty: 6</option>
            <option value="7">Qty: 7</option>
            <option value="8">Qty: 8</option>
            <option value="9">Qty: 9</option>
            <option value="10">Qty: 10</option>
          </select>

          <button type="submit" className="btn add-to-cart">
            Add to Cart
          </button>
        </form>

        <style jsx>{`
          form {
            display: flex;
            align-items: center;
          }
          .quantity {
            min-width: 5em;
          }
          .add-to-cart {
            flex: 1;
            margin-left: 1em;
          }
        `}</style>
      </div>
    );
  }
}

export const wrapCertificatePageController = (
  getDependencies: (ctx?: ClientContext) => ClientDependencies,
  renderContent: (ClientDependencies, ContentProps) => ?ReactElement<*>
) =>
  class CertificatePageController extends React.Component<InitialProps> {
    static async getInitialProps(ctx: ClientContext): Promise<InitialProps> {
      const { query: { id } } = ctx;
      const { deathCertificatesDao } = getDependencies(ctx);

      const certificate = await deathCertificatesDao.get(id);

      return {
        id,
        certificate,
      };
    }

    dependencies = getDependencies();

    addToCart = (quantity: number) => {
      const { cart } = this.dependencies;
      const { certificate } = this.props;

      if (certificate) {
        cart.add(certificate, quantity);
      }
    };

    render() {
      const { addToCart } = this;
      const { id, certificate } = this.props;

      return renderContent(this.dependencies, { id, certificate, addToCart });
    }
  };

export default wrapCertificatePageController(
  getDependencies,
  wrapAppLayout('checkout', (_, props) => <CertificatePageContent {...props} />)
);
