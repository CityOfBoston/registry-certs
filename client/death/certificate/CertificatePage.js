// @flow

import React, { type Element as ReactElement } from 'react';
import Head from 'next/head';
import Link from 'next/link';

import type { DeathCertificate } from '../../types';

import {
  getDependencies,
  type ClientContext,
  type ClientDependencies,
} from '../../app';

import AppLayout from '../../AppLayout';

import AddedToCartPopup from './AddedToCartPopup';

type InitialProps = {|
  id: string,
  certificate: ?DeathCertificate,
  backUrl: ?string,
|};

export type ContentProps = {
  ...InitialProps,
  addToCart: number => mixed,
  showAddedToCart: boolean,
  addedToCartQuantity: number,
  closeAddedToCart: () => mixed,
};

type ContentState = {
  quantity: ?number,
};

export class CertificatePageContent extends React.Component<
  ContentProps,
  ContentState
> {
  state = {
    quantity: 1,
  };

  handleQuantityChange = (ev: SyntheticInputEvent<*>) => {
    const { value } = ev.target;

    this.setState({
      quantity: value ? parseInt(value, 10) : null,
    });
  };

  handleAddToCart = (ev: SyntheticInputEvent<*>) => {
    ev.preventDefault();

    const { addToCart } = this.props;
    const { quantity } = this.state;

    if (!quantity) {
      return;
    }

    addToCart(quantity);
  };

  render() {
    const {
      id,
      certificate,
      backUrl,
      showAddedToCart,
      addedToCartQuantity,
      closeAddedToCart,
    } = this.props;

    return (
      <div>
        <Head>
          <title>Boston.gov — Death Certificate #{id}</title>
        </Head>

        <div className="p-a300">
          {backUrl && (
            <div className="m-b300">
              <Link href={backUrl}>
                <a style={{ fontStyle: 'italic' }}>← Back to search</a>
              </Link>
            </div>
          )}

          <div className="sh sh--b0">
            <h1 className="sh-title" style={{ marginBottom: 0 }}>
              Deceased Details
            </h1>
          </div>
        </div>

        <div className="p-a300 b--w">
          {certificate && this.renderCertificate(certificate)}
        </div>

        {showAddedToCart &&
          certificate && (
            <div className="md">
              <AddedToCartPopup
                certificate={certificate}
                quantity={addedToCartQuantity}
                close={closeAddedToCart}
              />
            </div>
          )}
      </div>
    );
  }

  renderCertificate({
    id,
    firstName,
    lastName,
    age,
    deathDate,
    deathYear,
    birthDate,
  }: DeathCertificate) {
    const { quantity } = this.state;

    return (
      <div>
        <ul className="dl">
          <li className="dl-i">
            <span className="dl-t">ID #</span>
            <span className="dl-d">{id}</span>
          </li>
          <li className="dl-i">
            <span className="dl-t">Full name</span>
            <span className="dl-d">
              {firstName} {lastName}
            </span>
          </li>
          {birthDate && (
            <li className="dl-i">
              <span className="dl-t">Date of birth</span>
              <span className="dl-d">{birthDate}</span>
            </li>
          )}
          <li className="dl-i">
            <span className="dl-t">Date of death</span>
            <span className="dl-d">{deathDate || deathYear}</span>
          </li>
          {age && (
            <li className="dl-i">
              <span className="dl-t">Age</span>
              <span className="dl-d">{age}</span>
            </li>
          )}

          <style jsx>{`
            .dl-i {
              display: flex;
              align-items: center;
              justify-content: flex-start;
            }
            .dl-t {
              padding-right: 1em;
            }
            .dl-t,
            .dl-d {
              line-height: 1rem;
              vertical-align: center;
            }
          `}</style>
        </ul>

        <form
          onSubmit={this.handleAddToCart}
          className="js-add-to-cart-form m-v500"
        >
          <input
            type="text"
            id="quantity"
            name="quantity"
            className="txt-f"
            value={quantity}
            onChange={this.handleQuantityChange}
          />

          <div className="sel-c sel-c--sq">
            <label htmlFor="quantity" className="a11y--h">
              Quantity:
            </label>

            <select
              name="quantityMenu"
              value={quantity}
              className="sel-f sel-f--sq"
              onChange={this.handleQuantityChange}
            >
              <option value="1">1</option>
              <option value="2">2</option>
              <option value="3">3</option>
              <option value="4">4</option>
              <option value="5">5</option>
              <option value="6">6</option>
              <option value="7">7</option>
              <option value="8">8</option>
              <option value="9">9</option>
              <option value="10">10</option>
            </select>
          </div>

          <button
            type="submit"
            className="btn add-to-cart"
            disabled={!quantity}
          >
            Add to Cart
          </button>

          <style jsx>{`
            form {
              display: flex;
              align-items: center;
            }

            .txt-f {
              /* matches the select box */
              height: 62px;
              width: 4em !important;
              text-align: right;
              border-right: none;
            }

            .sel-c:after {
              content: 'Qty.';
            }

            .add-to-cart {
              flex: 1;
              margin-left: 1em;
              height: 62px;
            }
          `}</style>
        </form>
      </div>
    );
  }
}

type ControllerState = {|
  showAddedToCart: boolean,
  addedToCartQuantity: number,
|};

export const wrapCertificatePageController = (
  getDependencies: (ctx?: ClientContext) => ClientDependencies,
  renderContent: (ClientDependencies, ContentProps) => ?ReactElement<*>
) =>
  class CertificatePageController extends React.Component<
    InitialProps,
    ControllerState
  > {
    static async getInitialProps(ctx: ClientContext): Promise<InitialProps> {
      const { query: { id, backUrl } } = ctx;
      const { deathCertificatesDao } = getDependencies(ctx);

      if (!id) {
        throw new Error('Missing id');
      }

      const certificate = await deathCertificatesDao.get(id);

      return {
        id,
        certificate,
        backUrl,
      };
    }

    dependencies = getDependencies();
    state: ControllerState = {
      showAddedToCart: false,
      addedToCartQuantity: 0,
    };

    addToCart = (quantity: number) => {
      const { cart } = this.dependencies;
      const { certificate } = this.props;

      if (certificate) {
        cart.add(certificate, quantity);

        this.setState({
          showAddedToCart: true,
          addedToCartQuantity: quantity,
        });
      }
    };

    closeAddedToCart = () => {
      this.setState({ showAddedToCart: false });
    };

    render() {
      const { addToCart, closeAddedToCart } = this;
      const { showAddedToCart, addedToCartQuantity } = this.state;

      return renderContent(this.dependencies, {
        ...this.props,
        addToCart,
        showAddedToCart,
        addedToCartQuantity,
        closeAddedToCart,
      });
    }
  };

export default wrapCertificatePageController(
  getDependencies,
  ({ cart }, props) => (
    <AppLayout navProps={{ cart, link: 'checkout' }}>
      <CertificatePageContent {...props} />
    </AppLayout>
  )
);
