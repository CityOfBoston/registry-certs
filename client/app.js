// @flow
/* global Stripe */

import { useStrict } from 'mobx';
import Router from 'next/router';

import type { Context as NextContext } from 'next';

import RouterListener from './lib/RouterListener';

import makeLoopbackGraphql, {
  type LoopbackGraphql,
} from './lib/loopback-graphql';

import type { RequestAdditions } from '../server/lib/request-additions';

import Cart from './store/Cart';
import OrderProvider from './store/OrderProvider';

import DeathCertificatesDao from './dao/DeathCertificatesDao';
import CheckoutDao from './dao/CheckoutDao';

export type ClientContext = NextContext<RequestAdditions>;

export type ClientDependencies = {
  stripe: ?StripeInstance,
  loopbackGraphql: LoopbackGraphql,
  cart: Cart,
  orderProvider: OrderProvider,
  deathCertificatesDao: DeathCertificatesDao,
  checkoutDao: CheckoutDao,
};

let browserInited = false;
let browserDependencies: ClientDependencies;

// Browser-only setup
export function initBrowser() {
  if (browserInited) {
    return;
  }

  browserInited = true;

  useStrict(true);

  const routerListener = new RouterListener();
  routerListener.attach(Router);
}

// Works on both server and browser. Memoizes on browser, so these dependencies
// will share state across pages.
//
// Typically called from a wrapper function or controller component to inject
// these into a component's props.
export function getDependencies(ctx?: ClientContext): ClientDependencies {
  if (process.browser && browserDependencies) {
    return browserDependencies;
  }

  // req will exist only when this function is called for getInitialProps on the
  // server.
  const req = ctx && ctx.req;

  const stripe =
    typeof Stripe !== 'undefined'
      ? Stripe((window.__NEXT_DATA__ || {}).stripePublishableKey || '')
      : null;
  const loopbackGraphql = makeLoopbackGraphql(req);
  const deathCertificatesDao = new DeathCertificatesDao(loopbackGraphql);
  const checkoutDao = new CheckoutDao(loopbackGraphql, stripe);
  const cart = new Cart();
  const orderProvider = new OrderProvider();

  if (process.browser) {
    cart.attach(window.localStorage, deathCertificatesDao);
    orderProvider.attach(window.localStorage);
  }

  const dependencies: ClientDependencies = {
    stripe,
    cart,
    orderProvider,
    deathCertificatesDao,
    checkoutDao,
    loopbackGraphql,
  };

  if (process.browser) {
    browserDependencies = dependencies;
  }

  return dependencies;
}
