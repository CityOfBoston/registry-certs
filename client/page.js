// @flow

import React from 'react';
import type { Context } from 'next';
import { useStrict } from 'mobx';
import Router from 'next/router';

import Cart from './store/Cart';
import makeLoopbackGraphql from './loopback-graphql';
import type { LoopbackGraphql } from './loopback-graphql';
import type { RequestAdditions } from '../server/request-additions';

import DeathCertificatesDao from './dao/DeathCertificatesDao';
import RouterListener from './RouterListener';

// Higher-order component for a Page in our app.

// Pass this mixin a function that requires the component class and returns it.
// require is necessary rather than import since imports are hoisted to the
// beginning of files, and we need to rehydrate before any Glamor "css"
// statements are processed.

export type ClientDependencies = {
  cart: Cart,
  deathCertificatesDao: DeathCertificatesDao,
  loopbackGraphql: LoopbackGraphql,
};

let browserInited = false;
let browserDependencies: ClientDependencies;

function maybeInitBrowserLibraries() {
  if (browserInited || !process.browser) {
    return;
  }

  browserInited = true;

  useStrict(true);

  const routerListener = new RouterListener();
  routerListener.attach(Router);
}

function makeDependencies(req: ?RequestAdditions): ClientDependencies {
  if (process.browser && browserDependencies) {
    return browserDependencies;
  }

  const loopbackGraphql = makeLoopbackGraphql(req);
  const deathCertificatesDao = new DeathCertificatesDao(loopbackGraphql);

  const cart = new Cart();
  if (process.browser) {
    cart.attach(window.localStorage, deathCertificatesDao);
  }

  const dependencies: ClientDependencies = {
    cart,
    deathCertificatesDao,
    loopbackGraphql,
  };

  if (process.browser) {
    browserDependencies = dependencies;
  }

  return dependencies;
}

export default <OP, P: $Subtype<Object>, S>(
  Component: Class<React.Component<OP, P, S>>
): Class<React.Component<void, P, void>> => {
  maybeInitBrowserLibraries();

  return class Page extends React.Component {
    props: P;
    loopbackGraphql: LoopbackGraphql = makeLoopbackGraphql();

    static getInitialProps(ctx: Context<*>) {
      const { req } = ctx;

      const dependencies = makeDependencies(req);

      if (typeof Component.getInitialProps === 'function') {
        return Component.getInitialProps(ctx, dependencies);
      } else {
        return {};
      }
    }

    render() {
      const dependencies = makeDependencies();
      return <Component {...dependencies} {...this.props} />;
    }
  };
};
