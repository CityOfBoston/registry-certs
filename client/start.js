// @flow

// Utility file for automatically calling initialization functions when page
// modules are run. We delegate to a separate module so that that module can be
// tested without automatically running on import.
//
// Note that this is called for every page on the browser, but only once on the
// server when the app is first interpreted on startup.
//
// This is typically imported and then the page component is require'd in. Using
// require enforces an execution order such that initBrowser can e.g. hydrate
// CSS-in-JS libraries before the "css" method is called during component module
// execution.

import { initBrowser } from './app.js';

if (process.browser) {
  initBrowser();
}
