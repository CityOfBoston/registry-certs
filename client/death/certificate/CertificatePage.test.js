// @flow
import React from 'react';
import { shallow } from 'enzyme';

import Cart from '../../store/Cart';
import DeathCertificatesDao from '../../dao/DeathCertificatesDao';

import CertificatePage, {
  wrapCertificatePageController,
  CertificatePageContent,
} from './CertificatePage';

import { TYPICAL_CERTIFICATE } from '../../../fixtures/client/death-certificates';

jest.mock('next/router');
jest.mock('../../dao/DeathCertificatesDao');

describe('integration', () => {
  it('renders', () => {
    expect(
      new CertificatePage({
        certificate: TYPICAL_CERTIFICATE,
        id: TYPICAL_CERTIFICATE.id,
        backUrl: '/search?q=jayne',
      }).render()
    ).toMatchSnapshot();
  });
});

describe('controller', () => {
  describe('getInitialProps', () => {
    let CertificatePageController;
    let deathCertificatesDao;

    beforeEach(() => {
      deathCertificatesDao = new DeathCertificatesDao((null: any));

      const dependencies: any = { deathCertificatesDao };
      CertificatePageController = wrapCertificatePageController(
        () => dependencies,
        () => null
      );
    });

    it('loads the cert passed in query', async () => {
      deathCertificatesDao.get.mockReturnValue(TYPICAL_CERTIFICATE);

      const initialProps = await CertificatePageController.getInitialProps(
        ({ query: { id: '000002' } }: any)
      );

      expect(deathCertificatesDao.get).toHaveBeenCalledWith('000002');
      expect(initialProps).toMatchSnapshot();
    });

    it('handles a 404', async () => {
      deathCertificatesDao.get.mockReturnValue(null);

      const initialProps = await CertificatePageController.getInitialProps(
        ({ query: { id: '000002' } }: any)
      );

      expect(deathCertificatesDao.get).toHaveBeenCalledWith('000002');
      expect(initialProps).toMatchSnapshot();
    });
  });

  describe('operations', () => {
    describe('setCartQuantity', () => {
      let cart;
      let controller;

      beforeEach(() => {
        cart = new Cart();

        const dependencies: any = { cart };
        const CertificatePageController = wrapCertificatePageController(
          () => dependencies,
          () => null
        );

        controller = shallow(
          <CertificatePageController
            id="0002"
            certificate={TYPICAL_CERTIFICATE}
            backUrl="/search?q=jayne"
          />
        ).instance();

        controller.setCartQuantity(5);
      });

      it('adds 5 things to the cart', () => {
        expect(cart.size).toEqual(5);
      });
    });
  });
});

describe('content', () => {
  let wrapper;
  let setCartQuantity;

  beforeEach(() => {
    setCartQuantity = jest.fn();

    wrapper = shallow(
      <CertificatePageContent
        certificate={TYPICAL_CERTIFICATE}
        id={TYPICAL_CERTIFICATE.id}
        backUrl={'/search?q=jayne'}
        setCartQuantity={setCartQuantity}
        cartQuantity={0}
      />
    );
  });

  it('changes quantity and adds to cart', () => {
    const form = wrapper.find('form');
    const quantityMenu = wrapper.find('[name="quantity"]');

    quantityMenu.simulate('change', { target: { value: '5' } });
    form.simulate('submit', { preventDefault: jest.fn() });

    expect(setCartQuantity).toHaveBeenCalledWith(5);
  });
});
