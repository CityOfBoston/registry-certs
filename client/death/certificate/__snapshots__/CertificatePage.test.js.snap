// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`controller getInitialProps handles a 404 1`] = `
Object {
  "backUrl": undefined,
  "certificate": null,
  "id": "000002",
}
`;

exports[`controller getInitialProps loads the cert passed in query 1`] = `
Object {
  "backUrl": undefined,
  "certificate": Object {
    "age": "53",
    "birthDate": "5/1/1962",
    "deathDate": "3/4/2016",
    "deathYear": "2016",
    "firstName": "BRUCE",
    "id": "000002",
    "lastName": "BANNER",
    "pending": false,
  },
  "id": "000002",
}
`;

exports[`integration renders 1`] = `
<AppLayout
  navProps={
    Object {
      "cart": Cart {},
      "link": "checkout",
    }
  }
>
  <CertificatePageContent
    addToCart={[Function]}
    addedToCartQuantity={0}
    backUrl="/search?q=jayne"
    certificate={
      Object {
        "age": "53",
        "birthDate": "5/1/1962",
        "deathDate": "3/4/2016",
        "deathYear": "2016",
        "firstName": "BRUCE",
        "id": "000002",
        "lastName": "BANNER",
        "pending": false,
      }
    }
    closeAddedToCart={[Function]}
    id="000002"
    showAddedToCart={false}
  />
</AppLayout>
`;
